packages/confirm-dialog/src/vaadin-confirm-dialog.js

```diff
@@ -268,73 +274,200 @@ class ConfirmDialog extends ElementMixin(ThemableMixin(PolymerElement)) {
     }
   }
 
+  static get observers() {
+    return [
+      '__updateConfirmButton(_confirmButton, confirmText, confirmTheme)',
+      '__updateCancelButton(_cancelButton, cancelText, cancelTheme, cancel)',
+      '__updateHeaderNode(_headerNode, header)',
+      '__updateMessageNode(_messageNode, message)',
+      '__updateRejectButton(_rejectButton, rejectText, rejectTheme, reject)'
+    ];
+  }
+
+  /** @protected */
+  get slots() {
+    // NOTE: order in which slots are listed matches the template.
+    return {
+      ...super.slots,
+      header: () => {
+        const h3 = document.createElement('h3');
+        this.__defaultHeader = h3;
+        return h3;
+      },
+      '': () => {
+        const div = document.createElement('div');
+        this.__defaultMessage = div;
+        return div;
+      },
+      'cancel-button': () => {
+        const button = document.createElement('vaadin-button');
+        button.setAttribute('theme', this.cancelTheme);
+        button.textContent = this.cancelText;
+        return button;
+      },
+      'reject-button': () => {
+        const button = document.createElement('vaadin-button');
+        button.setAttribute('theme', this.rejectTheme);
+        button.textContent = this.rejectText;
+        return button;
+      },
+      'confirm-button': () => {
+        return document.createElement('vaadin-button');
+      }
+    };
+  }
+
+  constructor() {
+    super();
+    this.__slottedNodes = [];
+    this._observer = new FlattenedNodesObserver(this, (info) => {
+      this.__onDomChange(info.addedNodes);
+    });
+  }
+
   /** @protected */
   ready() {
     super.ready();
 
-    this.$.dialog.$.overlay.addEventListener('vaadin-overlay-escape-press', this._escPressed.bind(this));
+    this.__boundCancel = this._cancel.bind(this);
+    this.__boundConfirm = this._confirm.bind(this);
+    this.__boundReject = this._reject.bind(this);
+
+    this._overlayElement = this.$.dialog.$.overlay;
+    this._overlayElement.addEventListener('vaadin-overlay-escape-press', this._escPressed.bind(this));
+    this._overlayElement.addEventListener('vaadin-overlay-open', () => this.__onDialogOpened());
+    this._overlayElement.addEventListener('vaadin-confirm-dialog-close', () => this.__onDialogClosed());
+
     if (this._dimensions) {
       Object.keys(this._dimensions).forEach((name) => {
         this._setDimension(name, this._dimensions[name]);
       });
     }
   }
 
-  /**
-   * @param {string} name
-   * @param {?string} oldValue
-   * @param {?string} newValue
-   * @protected
-   */
-  attributeChangedCallback(name, oldValue, newValue) {
-    super.attributeChangedCallback(name, oldValue, newValue);
-    if (name === 'dir') {
-      const value = newValue === 'rtl';
-      this.__isRTL = value;
-      this.opened && this.__toggleContentRTL(value);
+  /** @private */
+  __onDialogOpened() {
+    const overlay = this._overlayElement;
+
+    // Teleport slotted nodes to the overlay element.
```

Comment:
Why create the elements to light DOM of `<vaadin-confirm-dialog>` and teleport them to the `<vaadin-confirm-dialog-overlay>` on open? Wouldn't it work if `ConfirmDialogOverlay` instead extended `SlotMixin` so that the buttons would naturally exist in its light DOM?

Reply:
This was my first idea. I tried that and the main challenge with this approach was moving the nodes back to `vaadin-confirm-dialog`. So I decided to go with the same approach as in `vaadin-crud`.

Reply:
I was wondering why the buttons need to be inside `<vaadin-confirm-dialog>` at any point? With CRUD, the teleportation is necessary (?), but with confirm dialog, the buttons only even need to exist inside the `<vaadin-confirm-dialog-overlay>` or am I missing something?

Reply:
Ping @web-padawan

Reply:
Sorry, I missed your comment 😞   We currently expect user to add custom buttons before opening the confirm dialog.
IIRC there is an issue somewhere about not being able to add nodes while it’s opened.